File: packages/cordova-plugin-actito-scannables/src/ios/ActitoScannablesPlugin.swift
import ActitoKit
import ActitoScannablesKit

@objc(ActitoScannablesPlugin)
class ActitoScannablesPlugin : CDVPlugin {

    private var rootViewController: UIViewController? {
        get {
            UIApplication.shared.delegate?.window??.rootViewController
        }
    }

    override func pluginInitialize() {
        super.pluginInitialize()

        Actito.shared.scannables().delegate = self
    }

    @objc func registerListener(_ command: CDVInvokedUrlCommand) {
        ActitoScannablesPluginEventBroker.startListening(settings: commandDelegate.settings) { event in
            var payload: [String: Any] = [
                "name": event.name,
            ]

            if let data = event.payload {
                payload["data"] = data
            }

            let result = CDVPluginResult(status: .ok, messageAs: payload)
            result!.keepCallback = true

            self.commandDelegate!.send(result, callbackId: command.callbackId)
        }
    }

    // MARK: - Actito Scannables

    @objc func canStartNfcScannableSession(_ command: CDVInvokedUrlCommand) {
        let result = CDVPluginResult(status: .ok, messageAs: Actito.shared.scannables().canStartNfcScannableSession)
        self.commandDelegate!.send(result, callbackId: command.callbackId)
    }

    @objc func startScannableSession(_ command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            guard let rootViewController = self.rootViewController else {
                let result = CDVPluginResult(status: .error, messageAs: "Cannot start a scannable session with a nil root view controller.")
                self.commandDelegate!.send(result, callbackId: command.callbackId)

                return
            }

            Actito.shared.scannables().startScannableSession(controller: rootViewController)

            let result = CDVPluginResult(status: .ok)
            self.commandDelegate!.send(result, callbackId: command.callbackId)
        }
    }

    @objc func startNfcScannableSession(_ command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            Actito.shared.scannables().startNfcScannableSession()

            let result = CDVPluginResult(status: .ok)
            self.commandDelegate!.send(result, callbackId: command.callbackId)
        }
    }

    @objc func startQrCodeScannableSession(_ command: CDVInvokedUrlCommand) {
        DispatchQueue.main.async {
            guard let rootViewController = self.rootViewController else {
                let result = CDVPluginResult(status: .error, messageAs: "Cannot start a scannable session with a nil root view controller.")
                self.commandDelegate!.send(result, callbackId: command.callbackId)

                return
            }

            Actito.shared.scannables().startQrCodeScannableSession(controller: rootViewController, modal: true)

            let result = CDVPluginResult(status: .ok)
            self.commandDelegate!.send(result, callbackId: command.callbackId)
        }
    }

    @objc func fetch(_ command: CDVInvokedUrlCommand) {
        let tag = command.argument(at: 0) as! String

        Actito.shared.scannables().fetch(tag: tag) { result in
            switch result {
            case let .success(scannable):
                do {
                    let json = try scannable.toJson()

                    let result = CDVPluginResult(status: .ok, messageAs: json)
                    self.commandDelegate!.send(result, callbackId: command.callbackId)
                } catch {
                    let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                    self.commandDelegate!.send(result, callbackId: command.callbackId)
                }
            case let .failure(error):
                let result = CDVPluginResult(status: .error, messageAs: error.localizedDescription)
                self.commandDelegate!.send(result, callbackId: command.callbackId)
            }
        }
    }
}

extension ActitoScannablesPlugin: ActitoScannablesDelegate {
    func actito(_ actitoScannables: ActitoScannables, didDetectScannable scannable: ActitoScannable) {
        do {
            ActitoScannablesPluginEventBroker.dispatchEvent(
                name: "scannable_detected",
                payload: try scannable.toJson()
            )
        } catch {
            logger.error("Failed to emit the scannable_detected event.", error: error)
        }
    }

    func actito(_ actitoScannables: ActitoScannables, didInvalidateScannerSession error: Error) {
        ActitoScannablesPluginEventBroker.dispatchEvent(
            name: "scannable_session_failed",
            payload: error.localizedDescription
        )
    }
}

File: packages/cordova-plugin-actito-scannables/src/android/ActitoScannablesPlugin.kt
package com.actito.scannables.cordova

import org.apache.cordova.CallbackContext
import org.apache.cordova.CordovaArgs
import org.apache.cordova.CordovaPlugin
import org.apache.cordova.PluginResult
import org.json.JSONArray
import org.json.JSONObject
import com.actito.Actito
import com.actito.ActitoCallback
import com.actito.scannables.ActitoScannables
import com.actito.scannables.ktx.scannables
import com.actito.scannables.models.ActitoScannable

class ActitoScannablesPlugin : CordovaPlugin(), ActitoScannables.ScannableSessionListener {

    override fun pluginInitialize() {
        logger.hasDebugLoggingEnabled = Actito.options?.debugLoggingEnabled ?: false

        Actito.scannables().addListener(this)
    }

    override fun onDestroy() {
        Actito.scannables().removeListener(this)
    }

    override fun execute(action: String, args: CordovaArgs, callback: CallbackContext): Boolean {
        when (action) {
            "canStartNfcScannableSession" -> canStartNfcScannableSession(args, callback)
            "startScannableSession" -> startScannableSession(args, callback)
            "startNfcScannableSession" -> startNfcScannableSession(args, callback)
            "startQrCodeScannableSession" -> startQrCodeScannableSession(args, callback)
            "fetch" -> fetch(args, callback)

            // Event broker
            "registerListener" -> registerListener(args, callback)

            else -> {
                callback.error("No implementation for action '$action'.")
                return false
            }
        }

        return true
    }

    // region Actito Scannables

    private fun canStartNfcScannableSession(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        callback.success(Actito.scannables().canStartNfcScannableSession)
    }

    private fun startScannableSession(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        val activity = cordova.activity ?: run {
            callback.error("Cannot start a scannable session without an activity.")
            return
        }

        Actito.scannables().startScannableSession(activity)
        callback.void()
    }

    private fun startNfcScannableSession(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        val activity = cordova.activity ?: run {
            callback.error("Cannot start a scannable session without an activity.")
            return
        }

        Actito.scannables().startNfcScannableSession(activity)
        callback.void()
    }

    private fun startQrCodeScannableSession(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        val activity = cordova.activity ?: run {
            callback.error("Cannot start a scannable session without an activity.")
            return
        }

        Actito.scannables().startQrCodeScannableSession(activity)
        callback.void()
    }

    private fun fetch(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        val tag = args.getString(0)

        Actito.scannables().fetch(tag, object : ActitoCallback<ActitoScannable> {
            override fun onSuccess(result: ActitoScannable) {
                try {
                    callback.success(result.toJson())
                } catch (e: Exception) {
                    callback.error(e.message)
                }
            }

            override fun onFailure(e: Exception) {
                callback.error(e.message)
            }
        })
    }

    // endregion

    // region ActitoScannables.ScannableSessionListener

    override fun onScannableDetected(scannable: ActitoScannable) {
        try {
            ActitoScannablesPluginEventBroker.dispatchEvent("scannable_detected", scannable.toJson())
        } catch (e: Exception) {
            logger.error("Failed to emit the scannable_detected event.", e)
        }
    }

    override fun onScannableSessionError(error: Exception) {
        try {
            ActitoScannablesPluginEventBroker.dispatchEvent("scannable_session_failed", error.localizedMessage)
        } catch (e: Exception) {
            logger.error("Failed to emit the scannable_session_failed event.", e)
        }
    }

    // endregion

    private fun registerListener(@Suppress("UNUSED_PARAMETER") args: CordovaArgs, callback: CallbackContext) {
        ActitoScannablesPluginEventBroker.setup(preferences, object : ActitoScannablesPluginEventBroker.Consumer {
            override fun onEvent(event: ActitoScannablesPluginEventBroker.Event) {
                val payload = JSONObject()
                payload.put("name", event.name)
                when (event.payload) {
                    null -> {} // Skip encoding null payloads.
                    is Boolean -> payload.put("data", event.payload)
                    is Int -> payload.put("data", event.payload)
                    is Float -> payload.put("data", event.payload)
                    is Double -> payload.put("data", event.payload)
                    is String -> payload.put("data", event.payload)
                    is JSONObject -> payload.put("data", event.payload)
                    is JSONArray -> payload.put("data", event.payload)
                    else -> throw IllegalArgumentException("Unsupported event payload of type '${event.payload::class.java.simpleName}'.")
                }

                val result = PluginResult(PluginResult.Status.OK, payload)
                result.keepCallback = true

                callback.sendPluginResult(result)
            }
        })
    }
}

private fun CallbackContext.void() {
    sendPluginResult(PluginResult(PluginResult.Status.OK, null as String?))
}

private fun CallbackContext.success(b: Boolean) {
    sendPluginResult(PluginResult(PluginResult.Status.OK, b))
}
